src/Apple/Provider.php

```diff
@@ -0,0 +1,204 @@
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenUrl()
+    {
+        return self::URL.'/auth/token';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getCodeFields($state = null)
+    {
+        $fields = [
+            'client_id'     => $this->clientId,
+            'redirect_uri'  => $this->redirectUrl,
+            'scope'         => $this->formatScopes($this->getScopes(), $this->scopeSeparator),
+            'response_type' => 'code',
+            'response_mode' => 'form_post',
+        ];
+
+        if ($this->usesState()) {
+            $fields['state'] = $state;
+            $fields['nonce'] = Str::uuid().'-'.$state;
+        }
+
+        return array_merge($fields, $this->parameters);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAccessTokenResponse($code)
+    {
+        $response = $this->getHttpClient()->post($this->getTokenUrl(), [
+            'headers'        => ['Authorization' => 'Basic '.base64_encode($this->clientId.':'.$this->clientSecret)],
+            'form_params'    => $this->getTokenFields($code),
+        ]);
+
+        return json_decode($response->getBody(), true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenFields($code)
+    {
+        return array_merge(parent::getTokenFields($code), [
+            'grant_type' => 'authorization_code',
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getUserByToken($token)
+    {
+        static::verify($token);
+        $claims = explode('.', $token)[1];
+
+        return json_decode(base64_decode($claims), true);
+    }
+
+    /**
+     * Verify apple jwt.
+     *
+     * @param $jwt
+     *
+     * @return object
+     *
+     * @see https://appleid.apple.com/auth/keys
+     */
+    public static function verify($jwt)
+    {
+        $signer = new Sha256();
+
+        $token = (new Parser())->parse((string) $jwt);
+
+        if ($token->getClaim('iss') !== self::URL) {
+            throw new InvalidStateException('Invalid Issuer', Response::HTTP_UNAUTHORIZED);
+        }
+        if ($token->isExpired()) {
+            throw new InvalidStateException('Token Expired', Response::HTTP_UNAUTHORIZED);
+        }
+
+        $data = json_decode(file_get_contents(self::URL.'/auth/keys'), true);
+
+        $public_keys = JWK::parseKeySet($data);
+
+        $signature_verified = false;
+
+        foreach ($public_keys as $res) {
+            $publicKey = openssl_pkey_get_details($res);
+            if ($token->verify($signer, $publicKey['key'])) {
+                $signature_verified = true;
+            }
+        }
+        if (!$signature_verified) {
+            throw new InvalidStateException('Invalid JWT Signature');
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function user()
+    {
+        //Temporary fix to enable stateless
```

Comment:
Can you explain this?

Reply:
According to Apple, the valid scopes(name and email) are only available under the response mode `form_post`. So the Socialite Apple callback should be always pointed at a post route while doing this Apple callback is initiating the new session from Apple which is different from the current user session. We tried if this is a cors issue and it is not. Currently, the CSRF is disabled for the callback route. In order to support the stateful applications which depend on sessions, we wrote a workaround so that the session state variable is passed to apple and then taken back. This is a temporary workaround. Users can still validate the state by comparing the session state variable from the Apple session with the existing session.

Reply:
According to Apple, the valid scopes(name and email) are only available under the response mode form_post. So the Socialite Apple callback should be always pointed at a post route while doing this Apple callback is initiating the new session from Apple which is different from the current user session. We tried if this is a cors issue and it is not. Currently, the CSRF is disabled for the callback route. In order to support the stateful applications which depend on sessions, we wrote a workaround so that the session state variable is passed to apple and then taken back. This is a temporary workaround. Users can still validate the state by comparing the session state variable from the Apple session with the existing session.

Reply:
Currently, we have implemented state verification using the JWT nonce. Users can still verify the state by passing the session `state_verify`  variable.